### PR TITLE
Fix issue #4

### LINKED ISSUE #4: Finalizers are unreliable
Currently, the `log` package [relies on finalizers](https://github.com/scrive/log/blob/75f462382ea5278362a4222d5487adf4a01f747e/src/Log/Logger.hs#L83) to ensure that all messages in the logger's input queue have been serialised:

```haskell
    void . mkWeakIORef finalizer $ do
      atomically waitForWrite
      killThread tid
```

The idea is that when the `Logger` record gets garbage collected the RTS will run the associated finalizer, which will ensure that all outstanding messages in the logger's input queue have been serialised and kill the logger thread. The problem is that there is no guarantee that the finalizer will actually run! Quoting the [`System.Mem.Weak` documentation](https://hackage.haskell.org/package/base-4.9.0.0/docs/System-Mem-Weak.html#t:Weak):

> It is not guaranteed that a finalizer will eventually run, and no attempt is made to run outstanding finalizers when the program exits. Therefore finalizers should not be relied on to clean up resources - other methods (eg. exception handlers) should be employed, possibly in addition to finalizers.

Here's a test program that triggers this problem in practice:

```haskell
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Log
import Log.Backend.ElasticSearch

import System.Random

main :: IO ()
main = do
  let config = ElasticSearchConfig {
        esServer  = "http://localhost:9200",
        esIndex   = "logs",
        esMapping = "log"
        }
  logger <- elasticSearchLogger config randomIO
  runLogT "main" logger $ do
    logTrace_ "foo"
  -- waitForLogger logger
```

Unless you uncomment the last line, this code will never write any messages to the log. We should ditch finalizers in favour of exception handling.

### PR DESCRIPTION
**Pull request by 23Skidoo**
This still doesn't completely free the users from the need to think, but should make it harder to get messages silently dropped.

Resolves #4.

### CODE REVIEW

File: src/Log/Monad.hs [hunk 1]
```diff
@@ -43,16 +43,20 @@ newtype LogT m a = LogT { unLogT :: InnerLogT m a }
...
         -> LogT m a -- ^ The 'LogT' computation to run.
         -> m a
-runLogT component logger m = runReaderT (unLogT m) LoggerEnv {
+runLogT component logger m = (runReaderT (unLogT m) LoggerEnv {
   leLogger = logger
 , leComponent = component
 , leDomain = []
 , leData = []
-}
+}) `finally` (liftIO $ waitForLogger logger)
```
Comment (arybczak) [inline src/Log/Monad.hs]: This is too restrictive - generally you want to wait for serialization of logs only at exit and runLogT can be run anywhere in the application. If in such case you run this with logger that serializes messages periodically, runLogT will introduce unnecessary delay.

Comment (23Skidoo) [inline src/Log/Monad.hs]: The only fix then is to require clients to add `` `finally` waitForLogger logger`` calls to their code.

**Comment by arybczak** [comment]
To be fair, if you want to get rid of finalizers then the optimal solution IMO is to rewrite functions that construct loggers to be alloca-like, e.g. `withBulkStdoutLogger :: (Logger -> IO r) -> IO r`, or `withPgLogger :: ConnectionSourceM IO -> (Logger -> IO r) -> IO r`. Then you can guarantee that waitForLogger is always run when Logger goes out of scope. This won't prevent anyone from using Logger that was already finalized by returning it outside the scope in which it's valid, but that's fine as you can do the same with `alloca`, so simply mentioning this in the documentation should be sufficient.

**Comment by 23Skidoo** [comment]
@arybczak Thanks, that's a good idea, I'll implement it.

> This won't prevent anyone from using Logger that was already finalized by returning it outside the scope in which it's valid

This can sort of be prevented by marking the logger as closed and raising an exception when it's used outside of scope. There will still be the issue of main thread exiting before the child threads, but that's just something clients need to be aware of.

**Comment by 23Skidoo** [comment]
Implemented new logger creation API as suggested by @arybczak, but not logger invalidation (yet).

**Comment by 23Skidoo** [comment]
Logger invalidation now implemented.

**Comment by jonathanjouty** [comment]
[Followup with ElasticSearch logging](https://trello.com/c/frJCSXz6/516-followup-with-elasticsearch-logging)

**Comment by jonathanjouty** [comment]
[Followup with ElasticSearch logging](https://trello.com/c/frJCSXz6/516-followup-with-elasticsearch-logging)

**Comment by jonathanjouty** [comment]
I've had a quick look and Travis CI is happy, so I'm not going to dig deeper right now.

@arybczak did you want to review changes?
If not, @23Skidoo feel free to merge.

**Comment by 23Skidoo** [comment]
I think I'll merge this now and address @arybczak's potential future comments later if/when he decides to review.

**Pull request merged**